### PR TITLE
Rewrite doins by python (bug 624526).

### PR DESCRIPTION
**Pull request by zmedico**
doins is written in bash. However, specifically in case that
too many files are installed, it is very slow.
This CL rewrites the script in python for performance.

BUG=chromium:712659
TEST=time (./setup_board --forace && \
     ./build_package --withdev && \
     ./build_image --noenable_rootfs_verification test)
===Before===
real    21m35.445s
user    93m40.588s
sys     21m31.224s

===After===
real    17m30.106s
user    94m1.812s
sys     20m13.468s

Change-Id: Ib10f623961ba316753d58397cff5e72fbc343339
Reviewed-on: https://chromium-review.googlesource.com/559225
X-Chromium-Bug: 712659
X-Chromium-Bug-url: https://bugs.chromium.org/p/chromium/issues/detail?id=712659
X-Gentoo-Bug: 624526
X-Gentoo-Bug-url: https://bugs.gentoo.org/624526

### CODE REVIEW

File: bin/ebuild-helpers/doins [hunk 1]
```diff
@@ -44,116 +44,17 @@ if [[ ${INSDESTTREE#${ED}} != "${INSDESTTREE}" ]]; then
...
-			((success|=1))
-		else
-			((failed|=1))
-		fi
-	fi
-done
-rm -rf "$TMP"
-[[ $failed -ne 0 || $success -eq 0 ]] && { __helpers_die "${helper} failed"; exit 1; } || exit 0
+"${PORTAGE_PYTHON:-/usr/bin/python}" \
+	"${PORTAGE_BIN_PATH:-/usr/lib/portage/bin}"/doins.py \
```
Comment (mrueg) [inline bin/ebuild-helpers/doins]: Is /usr/lib/portage/bin a safe fallback? Shouldn't we check for it being unset instead?

File: bin/ebuild-helpers/doins [hunk 2]
```diff
@@ -44,116 +44,17 @@ if [[ ${INSDESTTREE#${ED}} != "${INSDESTTREE}" ]]; then
...
-	PRESERVE_SYMLINKS=n
+	SYMLINK_OPTION=''
 fi
 
-export TMP=$(mktemp -d "${T}/.doins_tmp_XXXXXX")
-# Use separate directories to avoid potential name collisions.
-mkdir -p "$TMP"/{1,2}
+# Use safe cwd, avoiding unsafe import for bug #469338.
+export __PORTAGE_HELPER_CWD=${PWD}
+cd "${PORTAGE_PYM_PATH:-/usr/lib/portage/pym}"
```
Comment (mrueg) [inline bin/ebuild-helpers/doins]: Is this path a safe fallback?

File: bin/doins.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+		if self._copy_xattr:
+			self._xattr_exclude = os.environ.get(
+				'PORTAGE_XATTR_EXCLUDE',
+				'security.* system.nfs4_acl')
+
+	def run(self, source, dest_dir):
+		"""Installs a file at |source| into |dest_dir| in process."""
+		dest = os.path.join(dest_dir, os.path.basename(source))
+		try:
+			# TODO: Consider to use portage.util.file_copy.copyfile
```
Comment (mrueg) [inline bin/doins.py]: Possible to get this TODO done before merging?

File: bin/doins.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+class _DirSubprocessInstallRunner(object):
+	"""Runs real 'install' command to install files or directories."""
+
+	def __init__(self, split_options):
+		self._split_options = split_options
+
+	def run(self, dest):
+		"""Installs a dir into |dest| by 'install' command."""
+		command = ['install', '-d'] + self._split_options + [dest]
+		subprocess.call(command)
```
Comment (mrueg) [inline bin/doins.py]: Could need some exception handling, if subprocess.call fails

File: bin/doins.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+		return u.pw_uid
+	return int(user)
+
+
+def _parse_mode(mode):
+	# "install"'s --mode option is complicated. So here is partially
+	# supported.
+	# In Python 3, the prefix of octal int must be '0o' rather than '0'.
+	# So, set base explicitly in that case.
+	return int(mode, 8 if re.search(r'^0[0-7]*$', mode) else 0)
```
Comment (mrueg) [inline bin/doins.py]: Split up into multiple lines for readability and consider compiling the regex

**Comment by zmedico** [comment]
Updated with the version merged in [chromiumos/third_party/portage_tool](https://chromium.googlesource.com/chromiumos/third_party/portage_tool/+/dd8ad7251ab85091ebcac11b721eb82e21d82256), and added a commit to optimize with copyfile from portage.util.file_copy.

**Pull request merged**